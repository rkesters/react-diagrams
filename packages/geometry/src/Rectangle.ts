import { isNumber } from 'lodash';
import { Point } from './Point';
import { Polygon } from './Polygon';

export class Rectangle extends Polygon {
	constructor(tl: Point, tr: Point, br: Point, bl: Point);
	constructor(position: Point, width: number, height: number);
	constructor(x?: number, y?: number, width?: number, height?: number);

	constructor(a: Point | number = 0, b: Point | number  = 0, c: Point | number  = 0, d: Point | number  = 0) {
		if (Point.isa(a) && Point.isa(b) && Point.isa(c) && Point.isa(d)) {
			super([a, b, c, d]);
			return;
		} else if (Point.isa(a) && isNumber(b) && isNumber(c) && isNumber(d)) {
			super([a, new Point(a.x + b, a.y), new Point(a.x + b, a.y + c), new Point(a.x, a.y + c)]);
			return;
		} else if (isNumber(a) && isNumber(b) && isNumber(c) && isNumber(d)) {
			super(Rectangle.pointsFromBounds(a, b, c, d));
			return;
		}

		throw new Error(`Rectangle call with invaild arguements: ${arguments.length}`);
	}

	static pointsFromBounds(x: number, y: number, width: number, height: number): Point[] {
		return [new Point(x, y), new Point(x + width, y), new Point(x + width, y + height), new Point(x, y + height)];
	}

	updateDimensions(x: number, y: number, width: number, height: number) {
		this.points = Rectangle.pointsFromBounds(x, y, width, height);
	}

	setPoints(points: Point[]) {
		if (points.length !== 4) {
			throw 'Rectangles must always have 4 points';
		}
		super.setPoints(points);
	}

	containsPoint(point: Point) {
		const tl = this.getTopLeft();
		const br = this.getBottomRight();

		return point.x >= tl.x && point.x <= br.x && point.y >= tl.y && point.y <= br.y;
	}

	getWidth(): number {
		return Math.sqrt(
			Math.pow(this.getTopLeft().x - this.getTopRight().x, 2) + Math.pow(this.getTopLeft().y - this.getTopRight().y, 2)
		);
	}

	getHeight(): number {
		return Math.sqrt(
			Math.pow(this.getBottomLeft().x - this.getTopLeft().x, 2) +
				Math.pow(this.getBottomLeft().y - this.getTopLeft().y, 2)
		);
	}

	getTopMiddle(): Point {
		return Point.middlePoint(this.getTopLeft(), this.getTopRight());
	}

	getBottomMiddle(): Point {
		return Point.middlePoint(this.getBottomLeft(), this.getBottomRight());
	}

	getLeftMiddle(): Point {
		return Point.middlePoint(this.getBottomLeft(), this.getTopLeft());
	}

	getRightMiddle(): Point {
		return Point.middlePoint(this.getBottomRight(), this.getTopRight());
	}

	getTopLeft(): Point {
		return this.points[0];
	}

	getTopRight(): Point {
		return this.points[1];
	}

	getBottomRight(): Point {
		return this.points[2];
	}

	getBottomLeft(): Point {
		return this.points[3];
	}
}
