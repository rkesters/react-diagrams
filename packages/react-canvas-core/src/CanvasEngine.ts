import { debounce } from 'lodash';
import { CanvasModel } from './entities/canvas/CanvasModel';
import { FactoryBank } from './core/FactoryBank';
import { AbstractReactFactory } from './core/AbstractReactFactory';
import { LayerModel } from './entities/layer/LayerModel';
import { BaseListener, BaseObserver } from './core/BaseObserver';
import { MouseEvent } from 'react';
import { BaseModel } from './core-models/BaseModel';
import { Point } from '@projectstorm/geometry';
import { ActionEventBus } from './core-actions/ActionEventBus';
import { ZoomCanvasAction } from './actions/ZoomCanvasAction';
import { DeleteItemsAction } from './actions/DeleteItemsAction';
import { StateMachine } from './core-state/StateMachine';
export interface CanvasEngineListener extends BaseListener {
	canvasReady(): void;

	repaintCanvas(): void;

	rendered(): void;
}

/**
 * Defines the CanvasEngine options
 */
export interface CanvasEngineOptions {
	registerDefaultDeleteItemsAction?: boolean;
	registerDefaultZoomCanvasAction?: boolean;
	/**
	 * Defines the debounce wait time in milliseconds if > 0
	 */
	repaintDebounceMs?: number;
}

export class CanvasEngine<
	L extends CanvasEngineListener = CanvasEngineListener,
	M extends CanvasModel = CanvasModel
> extends BaseObserver<L> {
	protected model: M ;
	protected layerFactories: FactoryBank<AbstractReactFactory<LayerModel>>;
	protected canvas: HTMLDivElement | undefined;
	protected eventBus: ActionEventBus;
	protected stateMachine: StateMachine;
	protected options: CanvasEngineOptions;

	constructor(model: M, options: CanvasEngineOptions = {}) {
		super();
		this.model = model;
		this.eventBus = new ActionEventBus(this);
		this.stateMachine = new StateMachine(this);
		this.layerFactories = new FactoryBank();
		this.registerFactoryBank(this.layerFactories);

		/**
		 * Overrides the standard options with the possible given options
		 */
		this.options = {
			registerDefaultDeleteItemsAction: true,
			registerDefaultZoomCanvasAction: true,
			repaintDebounceMs: 0,
			...options
		};
		if (this.options.registerDefaultZoomCanvasAction === true) {
			this.eventBus.registerAction(new ZoomCanvasAction({}, this));
		}
		if (this.options.registerDefaultDeleteItemsAction === true) {
			this.eventBus.registerAction(new DeleteItemsAction({}, this));
		}
	}

	getStateMachine() {
		return this.stateMachine;
	}

	getRelativeMousePoint(event: { clientX: number; clientY: number }): Point {
		const point = this.getRelativePoint(event.clientX, event.clientY);
		return new Point(
			(point.x - (this.model?.getOffsetX() ?? 0)) / ((this.model?.getZoomLevel() ?? 1) / 100.0),
			(point.y - (this.model?.getOffsetY() ?? 0)) / ((this.model?.getZoomLevel() ?? 1) / 100.0)
		);
	}

	getRelativePoint(x:number, y: number): Point {
		const canvasRect = this.canvas?.getBoundingClientRect();
		if (!canvasRect) {
			throw new Error(`Can't get relative point because canvas has not been defined.`)
		}
		return new Point(x - canvasRect.left, y - canvasRect.top);
	}

	registerFactoryBank(factory: FactoryBank<any>) {
		factory.registerListener({
			factoryAdded: (event) => {
				event.factory.setDiagramEngine(this);
			},
			factoryRemoved: (event) => {
				event.factory.setDiagramEngine(null);
			}, eventWillFire: ():void => {} , eventDidFire: ():void => {}
		});
	}

	getActionEventBus() {
		return this.eventBus;
	}

	getLayerFactories() {
		return this.layerFactories;
	}

	getFactoryForLayer<F extends AbstractReactFactory<LayerModel>>(layer: LayerModel | string | undefined) {
		if (!layer) { return;}
		if (typeof layer === 'string') {
			return this.layerFactories.getFactory(layer);
		}
		return this.layerFactories.getFactory(layer.getType());
	}

	setModel(model: M) {
		this.model = model;
		if (this.canvas) {
			requestAnimationFrame(() => {
				this.repaintCanvas();
			});
		}
	}

	getModel(): M {
		return this.model;
	}

	repaintCanvas(promise: true): Promise<any>;
	repaintCanvas(): void;
	repaintCanvas(promise?: true): Promise<any> | void {
		const repaintDebounceMs  = this.options.repaintDebounceMs ?? 0;

		/**
		 * The actual repaint function
		 */
		const repaint = () => {
			this.iterateListeners((listener) => {
				if (listener.repaintCanvas) {
					listener.repaintCanvas();
				}
			});
		};

		// if the `repaintDebounceMs` option is > 0, then apply the debounce
		let repaintFn = repaint;

		if (repaintDebounceMs > 0) {
			repaintFn = debounce(repaint, repaintDebounceMs);
		}

		if (promise) {
			return new Promise<void>((resolve) => {
				const l = this.registerListener({
					rendered: () => {
						resolve();
						l.deregister();
					}
				} as L);
				repaintFn();
			});
		}

		repaintFn();
	}

	setCanvas(canvas?: HTMLDivElement | null) {
		if (this.canvas !== canvas) {
			this.canvas = canvas ?? undefined;
			if (canvas) {
				this.fireEvent(undefined, 'canvasReady');
			}
		}
	}

	getCanvas() {
		return this.canvas;
	}

	getMouseElement(event: MouseEvent): BaseModel<any> | null | undefined  {
		return null;
	}

	zoomToFit() {
		if (!this.canvas) {
			throw new Error(`Can not zoom to fit because canvas is not defined`)
		}
		const xFactor = this.canvas.clientWidth / this.canvas.scrollWidth;
		const yFactor = this.canvas.clientHeight / this.canvas.scrollHeight;
		const zoomFactor = xFactor < yFactor ? xFactor : yFactor;

		this.model?.setZoomLevel((this.model?.getZoomLevel() ?? 1) * zoomFactor);
		this.model?.setOffset(0, 0);
		this.repaintCanvas();
	}
}
