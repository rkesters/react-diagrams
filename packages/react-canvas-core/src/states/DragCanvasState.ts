import { CanvasEngine } from '..';
import { AbstractDisplacementState, AbstractDisplacementStateEvent } from '../core-state/AbstractDisplacementState';
import { State } from '../core-state/State';

export interface DragCanvasStateOptions {
	/**
	 * If enabled, the canvas is available to drag
	 */
	allowDrag?: boolean;
}

export class DragCanvasState extends AbstractDisplacementState {
	// store this as we drag the canvas
	initialCanvasX: number ;
	initialCanvasY: number ;
	config: DragCanvasStateOptions;

	constructor(engine: CanvasEngine, options: DragCanvasStateOptions = {}) {
		super(engine,{
			name: 'drag-canvas'
		});
		this.config = {
			allowDrag: true,
			...options
		};
		this.initialCanvasX = 0;
		this.initialCanvasY= 0;
	}

	async activated(prev: DragCanvasState) {
		super.activated(prev);
		this.engine.getModel().clearSelection();
		await this.engine.repaintCanvas(true);

		// we can block layer rendering because we are only targeting the transforms
		for (let layer of this.engine.getModel().getLayers()) {
			layer.allowRepaint(false);
		}

		this.initialCanvasX = this.engine.getModel().getOffsetX() ?? 0;
		this.initialCanvasY = this.engine.getModel().getOffsetY() ?? 0;
	}

	deactivated(next: State) {
		super.deactivated(next);
		for (let layer of this.engine.getModel().getLayers()) {
			layer.allowRepaint(true);
		}
	}

	fireMouseMoved(event: AbstractDisplacementStateEvent) {
		if (this.config.allowDrag) {
			this.engine
				.getModel()
				.setOffset(this.initialCanvasX + event.displacementX, this.initialCanvasY + event.displacementY);
			this.engine.repaintCanvas();
		}
	}
}
