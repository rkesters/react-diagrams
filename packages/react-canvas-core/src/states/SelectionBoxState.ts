import { MouseEvent, TouchEvent } from 'react';
import { AbstractDisplacementState, AbstractDisplacementStateEvent } from '../core-state/AbstractDisplacementState';
import { State } from '../core-state/State';
import { SelectionLayerModel } from '../entities/selection/SelectionLayerModel';
import { Point, Rectangle } from '@projectstorm/geometry';
import { ModelGeometryInterface } from '../core/ModelGeometryInterface';
import { CanvasEngine } from '..';

export interface IClientRect {
	left: number;
	top: number;
	right: number;
	bottom: number;
	width: number;
	height: number;
}

function isaMouseEvent(value: unknown): value is MouseEvent {
	return value instanceof MouseEvent
}

export class SelectionBoxState extends AbstractDisplacementState {
	layer: SelectionLayerModel ;

	constructor(engine: CanvasEngine) {
		super(engine, {
			name: 'selection-box'
		});
		this.layer = new SelectionLayerModel();
	}

	activated(previous: State) {
		super.activated(previous);
		this.engine.getModel().addLayer(this.layer);
	}

	deactivated(next: State) {
		super.deactivated(next);
		this.layer.remove();
		this.engine.repaintCanvas();
		this.layer = new SelectionLayerModel();
	}

	getBoxDimensions(stateEvent: AbstractDisplacementStateEvent): IClientRect {
		const event = stateEvent.event;

		const rel: Point = isaMouseEvent(event)  ? this.engine.getRelativePoint(event.clientX, event.clientY) :
		this.engine.getRelativePoint(event.touches[0].clientX, event.touches[0].clientY);


		return {
			left: rel.x > this.initialXRelative ? this.initialXRelative : rel.x,
			top: rel.y > this.initialYRelative ? this.initialYRelative : rel.y,
			width: Math.abs(rel.x - this.initialXRelative),
			height: Math.abs(rel.y - this.initialYRelative),
			right: rel.x < this.initialXRelative ? this.initialXRelative : rel.x,
			bottom: rel.y < this.initialYRelative ? this.initialYRelative : rel.y
		};
	}

	fireMouseMoved(event: AbstractDisplacementStateEvent) {
		this.layer.setBox(this.getBoxDimensions(event));

		const relative = this.engine.getRelativeMousePoint({
			clientX: this.initialX,
			clientY: this.initialY
		});
		if (event.virtualDisplacementX < 0) {
			relative.x -= Math.abs(event.virtualDisplacementX);
		}
		if (event.virtualDisplacementY < 0) {
			relative.y -= Math.abs(event.virtualDisplacementY);
		}
		const rect = new Rectangle(relative, Math.abs(event.virtualDisplacementX), Math.abs(event.virtualDisplacementY));

		for (let model of this.engine.getModel().getSelectionEntities()) {
			if ((model as unknown as ModelGeometryInterface).getBoundingBox) {
				const bounds = (model as unknown as ModelGeometryInterface).getBoundingBox();
				if (rect.containsPoint(bounds.getTopLeft()) && rect.containsPoint(bounds.getBottomRight())) {
					model.setSelected(true);
				} else {
					model.setSelected(false);
				}
			}
		}

		this.engine.repaintCanvas();
	}
}
