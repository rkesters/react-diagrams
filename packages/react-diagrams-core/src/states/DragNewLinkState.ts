import {
	AbstractDisplacementState,
	AbstractDisplacementStateEvent,
	Action,
	ActionEvent,
	InputType
} from '@projectstorm/react-canvas-core';
import { PortModel } from '../entities/port/PortModel';
import { MouseEvent } from 'react';
import { LinkModel } from '../entities/link/LinkModel';
import { DiagramEngine } from '../DiagramEngine';

export interface DragNewLinkStateOptions {
	/**
	 * If enabled, the links will stay on the canvas if they dont connect to a port
	 * when dragging finishes
	 */
	allowLooseLinks?: boolean;
	/**
	 * If enabled, then a link can still be drawn from the port even if it is locked
	 */
	allowLinksFromLockedPorts?: boolean;
}

export class DragNewLinkState extends AbstractDisplacementState<DiagramEngine> {
	port: PortModel | undefined | null;
	link: LinkModel | undefined | null;
	config: DragNewLinkStateOptions;

	constructor(engine: DiagramEngine, options: DragNewLinkStateOptions = {}) {
		super(engine, { name: 'drag-new-link' });

		this.config = {
			allowLooseLinks: true,
			allowLinksFromLockedPorts: false,
			...options
		};
		console.log(`DragNewLinkState registing actions`);
		this.registerAction(
			new Action({
				type: InputType.MOUSE_DOWN,
				//TODO: FIX ANY
				fire: (event: ActionEvent<MouseEvent, any>) => {
					console.log(`DragNewLinkState firing mouse down`);
					this.port = this.engine.getMouseElement(event.event) as PortModel;
					if (!this.config.allowLinksFromLockedPorts && this.port.isLocked()) {
						this.eject();
						return;
					}
					this.link = this.port.createLinkModel();

					// if no link is given, just eject the state
					if (!this.link) {
						console.log(`DragNewLinkState ejecting link`);
						this.eject();
						return;
					}
					this.link.setSelected(true);
					this.link.setSourcePort(this.port);
					this.engine.getModel().addLink(this.link);
					this.port.reportPosition();
				}
			},engine)
		);

		this.registerAction(
			new Action({
				type: InputType.MOUSE_UP,
				fire: (event: ActionEvent<MouseEvent>) => {
					const model = this.engine.getMouseElement(event.event);
					// check to see if we connected to a new port
					if (model instanceof PortModel) {
						if (this.port?.canLinkToPort(model)) {
							this.link?.setTargetPort(model);
							model.reportPosition();
							this.engine.repaintCanvas();
							return;
						} else {
							this.link?.remove();
							this.engine.repaintCanvas();
							return;
						}
					}

					if (!this.config.allowLooseLinks) {
						this.link?.remove();
						this.engine.repaintCanvas();
					}
				}
			}, engine)
		);
	}

	/**
	 * Calculates the link's far-end point position on mouse move.
	 * In order to be as precise as possible the mouse initialXRelative & initialYRelative are taken into account as well
	 * as the possible engine offset
	 */
	fireMouseMoved(event: AbstractDisplacementStateEvent):void {

		const portPos = this.port?.getPosition();
		if (!portPos) {return;}
		const zoomLevelPercentage = this.engine.getModel().getZoomLevel() / 100;
		const engineOffsetX = this.engine.getModel().getOffsetX() / zoomLevelPercentage;
		const engineOffsetY = this.engine.getModel().getOffsetY() / zoomLevelPercentage;
		const initialXRelative = this.initialXRelative / zoomLevelPercentage;
		const initialYRelative = this.initialYRelative / zoomLevelPercentage;
		const linkNextX = portPos.x - engineOffsetX + (initialXRelative - portPos.x) + event.virtualDisplacementX;
		const linkNextY = portPos.y - engineOffsetY + (initialYRelative - portPos.y) + event.virtualDisplacementY;

		this.link?.getLastPoint().setPosition(linkNextX, linkNextY);
		this.engine.repaintCanvas();
	}
}
