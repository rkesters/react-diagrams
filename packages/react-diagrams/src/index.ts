import {
	DefaultDiagramState,
	DiagramEngine,
	DiagramModel,
	LinkLayerFactory,
	NodeLayerFactory
} from '@projectstorm/react-diagrams-core';
import {
	DefaultLabelFactory,
	DefaultLinkFactory,
	DefaultNodeFactory,
	DefaultPortFactory
} from '@projectstorm/react-diagrams-defaults';
import { PathFindingLinkFactory } from '@projectstorm/react-diagrams-routing';
import { SelectionBoxLayerFactory, CanvasEngineOptions } from '@projectstorm/react-canvas-core';

export * from '@projectstorm/react-diagrams-core';
export * from '@projectstorm/react-diagrams-defaults';
export * from '@projectstorm/react-diagrams-routing';

/**
 * Construct an engine with the defaults installed
 */
export default (model: DiagramModel, options: CanvasEngineOptions = {}): DiagramEngine => {
	const engine = new DiagramEngine(model,options);

	// register model factories
	engine.getLayerFactories().registerFactory(new NodeLayerFactory(engine));
	engine.getLayerFactories().registerFactory(new LinkLayerFactory(engine));
	engine.getLayerFactories().registerFactory(new SelectionBoxLayerFactory(engine));

	engine.getLabelFactories().registerFactory(new DefaultLabelFactory(engine));
	engine.getNodeFactories().registerFactory(new DefaultNodeFactory(engine)); // i cant figure out why
	engine.getLinkFactories().registerFactory(new DefaultLinkFactory(engine));
	engine.getLinkFactories().registerFactory(new PathFindingLinkFactory(engine));
	engine.getPortFactories().registerFactory(new DefaultPortFactory(engine));

	// register the default interaction behaviours
	engine.getStateMachine().pushState(new DefaultDiagramState(engine));
	return engine;
};
