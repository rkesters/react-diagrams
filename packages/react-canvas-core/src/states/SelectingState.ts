import { State } from '../core-state/State';
import { Action, ActionEvent, InputType } from '../core-actions/Action';
import { SelectionBoxState } from './SelectionBoxState';
import { MouseEvent } from 'react';
import { CanvasEngine } from '..';

export class SelectingState extends State {
	constructor(engine: CanvasEngine) {
		super(engine, {
			name: 'selecting'
		});
		this.keys = ['shift'];

		this.registerAction(
			new Action({
				type: InputType.MOUSE_DOWN,
				fire: (event: ActionEvent<MouseEvent>) => {
					const element = this.engine.getActionEventBus().getModelForEvent(event);

					// go into a selection box on the canvas state
					if (!element) {
						this.transitionWithEvent(new SelectionBoxState(this.engine), event);
					} else {
						element.setSelected(true);
						this.engine.repaintCanvas();
					}
				}
			}, engine)
		);
	}
}
