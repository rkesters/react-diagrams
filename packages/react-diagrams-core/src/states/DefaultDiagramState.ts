import { MouseEvent, TouchEvent } from 'react';
import {
	SelectingState,
	State,
	Action,
	InputType,
	ActionEvent,
	DragCanvasState
} from '@projectstorm/react-canvas-core';
import { PortModel } from '../entities/port/PortModel';
import { DragNewLinkState } from './DragNewLinkState';
import { DiagramEngine } from '../DiagramEngine';
import { DragDiagramItemsState } from './DragDiagramItemsState';

export class DefaultDiagramState extends State<DiagramEngine> {
	dragCanvas: DragCanvasState;
	dragNewLink: DragNewLinkState;
	dragItems: DragDiagramItemsState;

	constructor(engine: DiagramEngine) {
		super(engine, {
			name: 'default-diagrams'
		});
		this.childStates = [new SelectingState(engine)];
		this.dragCanvas = new DragCanvasState(engine);
		this.dragNewLink = new DragNewLinkState(engine);
		this.dragItems = new DragDiagramItemsState(engine);

		// determine what was clicked on
		this.registerAction(
			new Action({
				type: InputType.MOUSE_DOWN,
				fire: (event: ActionEvent<MouseEvent>) => {
					const element = this.engine.getActionEventBus().getModelForEvent(event);

					// the canvas was clicked on, transition to the dragging canvas state
					if (!element) {
						this.transitionWithEvent(this.dragCanvas, event);
					}
					// initiate dragging a new link
					else if (element instanceof PortModel) {
						this.transitionWithEvent(this.dragNewLink, event);
					}
					// move the items (and potentially link points)
					else {
						this.transitionWithEvent(this.dragItems, event);
					}
				}
			},engine)
		);

		// touch drags the canvas
		this.registerAction(
			new Action({
				type: InputType.TOUCH_START,
				fire: (event: ActionEvent<TouchEvent>) => {
					this.transitionWithEvent(this.dragCanvas, event);
				}
			},engine)
		);
	}
}
