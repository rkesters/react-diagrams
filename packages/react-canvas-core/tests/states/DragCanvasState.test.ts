import { mocked } from 'ts-jest/utils';
import { CanvasEngine } from '../../src/CanvasEngine';
import { CanvasModel } from '../../src/entities/canvas/CanvasModel';
import { DragCanvasState } from '../../src/states/DragCanvasState';
import { TestLayer } from '../support/impls/TestLayer';



describe('DragCanvasState', () => {
	let engine = new CanvasEngine();
	beforeEach(() => {
		engine = new CanvasEngine();
		engine.setModel(new CanvasModel());
	});

	test('Constructor', () => {
		const state = new DragCanvasState();

		expect(state.getOptions()).toMatchSnapshot();
	});

	test('Activate', async () => {
		const state = new DragCanvasState();
		state.setEngine(engine);
		engine.getModel().setOffset(10, 20);
		const layer = new TestLayer();
		engine.getModel().addLayer(layer);
		engine.repaintCanvas = jest.fn();

		layer.allowRepaint = jest.fn();
		await state.activated({} as any);
		expect(state.initialCanvasX).toEqual(10);
		expect(state.initialCanvasY).toEqual(20);
		expect(layer.allowRepaint).toHaveBeenCalledWith(false);

		state.deactivated({} as any);
		expect(layer.allowRepaint).toHaveBeenCalledWith(true);

	});

	test('Fire Mouse Event', async () => {
		const state = new DragCanvasState();
		state.setEngine(engine);
		engine.getModel().setOffset(10, 20);
		const layer = new TestLayer();
		engine.getModel().addLayer(layer);
		engine.repaintCanvas = jest.fn();

		state.fireMouseMoved({displacementX: 30,displacementY: 50} as any);
		expect(engine.getModel().getOffsetX()).toEqual(state.initialCanvasX + 30);
		expect(engine.getModel().getOffsetY()).toEqual(state.initialCanvasY + 50);
		expect(engine.repaintCanvas).toHaveBeenCalled();

		state.config.allowDrag = false;
		mocked(engine.repaintCanvas).mockClear();
		state.fireMouseMoved({displacementX: 30,displacementY: 50} as any);

		expect(engine.repaintCanvas).not.toHaveBeenCalled();

	});

});
