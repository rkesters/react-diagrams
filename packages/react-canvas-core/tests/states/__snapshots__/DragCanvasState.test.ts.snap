// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragCanvasState Constructor 1`] = `
Object {
  "name": "drag-canvas",
}
`;
